Document destination diagrams

pierrechevalier83 opened the pull request.

pierrechevalier83 committed:
Document diagram for section member

pierrechevalier83 committed:
Document processing of incoming rellocation candidate

The little css change allows for 2 buttons on 2 consecutive lines to be
slightly spaced so they don't look like a single button.

pierrechevalier83 committed:
Document resource proof from a destination section's pov

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -102,6 +102,15 @@ <h2>Node's entire lifetime</h2>
   <h2>Node as a member of a section</h2>
+  <p>Once a node has joined a section, they need to be ready to take on multiple roles simultaneously:<br>
+  <ul>
+    <li> Deal with rellocation:<br>
```
typo. "relocation"
Multiple places, same for rellocate

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -102,6 +102,15 @@ <h2>Node's entire lifetime</h2>
+         The section as a whole will sometimes act as the source and sometimes as the destination of rellocations. This means nodes will sometimes rellocate away from this section and sometimes rellocate to this section. Nodes in this section will need to perform the right flow to deal with both of these situations.
+    <li> Decide who's an elder or a plain old adult
+  </ul>
+  All of these flows are happening simultaneously, but really, only one flow is active at anytime and all others must be in a "wait" state.<br>
```
I am not sure it is clear that for example, a node can handle relocating a node to its section at the same time it is relocating a node away from its section.

pierrechevalier83 replied:
Uhm. I wasn't 100% clear on this, but I thought that's what you meant. So what about:
> All of these flows are happening simultaneously, but in most cases, only one flow is active at anytime while the others are in a "wait" state.

Or would you just remove the entire sentence?

jeanphilippeD replied:
> All of these flows are happening simultaneously, but in most cases, only one flow is active at anytime while the others are in a "wait" state.

No, in all case, only one flow is active at a time, but there are multiple events to see and wait for in a flow, so you may have sent a RelocateResponse and are resource proofing, but in the wait, you may see the relocation trigger and send yourself an ExpectCandidate.


I'm not sure I would remove the sentence, the original sentence was not wrong.

jeanphilippeD replied:
I'm happy with the sentence as it is.

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
+  </p>
+  <button class="collapsible">RPC: ExpectCandidate</button>
+  <div class="content">
+    <p>We receive this RPC from a node that wants to join our section.<br>
```
We receive this RPC from a section that want to relocate a node to our section.

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
+  <button class="collapsible">RPC: ExpectCandidate</button>
+  <div class="content">
+    <p>We receive this RPC from a node that wants to join our section.<br>
+    They aren't connected to us, yet; so the RPC reaches us through their proxy node.<br>
```
The node is not communicating with us yet, only once we sent RelocateResponse RPC to the originating section.

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
+    <p>We receive this RPC from a node that wants to join our section.<br>
+    They aren't connected to us, yet; so the RPC reaches us through their proxy node.<br>
+    On receiving it, we vote for ParsecExpectCandidate to process it in the same order as other members of our section.<br>
+    It kickstarts the entire chain of events in this diagram.
```
> It kickstarts the entire chain of events in this diagram

...even if only enough node in our section received that event.

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
+  </div>
+  <button class="collapsible">shorter_prefix_section</button>
+  <div class="content">
+    <p>If we know of a section that has a shorter prefix than ours, we prefer for them to receive this incoming node rather than ourselves as it will help kepp the Network's sections tree balanced.<br>
```
kepp -> keep

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
+  <div class="content">
+    <p>If we know of a section that has a shorter prefix than ours, we prefer for them to receive this incoming node rather than ourselves as it will help kepp the Network's sections tree balanced.<br>
+    This shorter_prefix_section is a function that will return None if we are the shortest of any section we know, Some if there is a better candidate.<br>
+    If it is Some, we will relay the RPC to them instead of ourselves.
```
unclear, we do not relay RPC to ourself

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
+  </div>
+  <button class="collapsible">PROCESSING_CANDIDATE</button>
+  <div class="content">
+    <p>We want to allow at most one node at a time to enter our section through rellocation.<br>
```
We want to relocate at most one node at a time to our section.

pierrechevalier83 replied:
Uhm. I was trying to emphasize the "incoming" aspect... But I guess yours reads better...
Maybe some middle-ground:
> We want to accept only one incoming relocation at a time into our section.

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
+  <button class="collapsible">PROCESSING_CANDIDATE</button>
+  <div class="content">
+    <p>We want to allow at most one node at a time to enter our section through rellocation.<br>
+    The local variable: PROCESSING_CANDIDATE allows us to keep track of the fact someone is already passing the resource proof process, so we can't accept anyone else for now.<br>
```
> allows us to keep track of the fact someone

=> indicates if someone

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -132,8 +141,58 @@ <h2>Node as a member of a section</h2>
+  <div class="content">
+    <p>We want to allow at most one node at a time to enter our section through rellocation.<br>
+    The local variable: PROCESSING_CANDIDATE allows us to keep track of the fact someone is already passing the resource proof process, so we can't accept anyone else for now.<br>
+    When PROCESSING_CANDIDATE is true and we reach consensus on ParsecExpectCandidate, we simply send a RefuseCandidate RPC to the would-be-incoming-node so they know they must try another section or try again later.
```
somewhat inaccurate:
- only refuse if we are the shortest_prefix.
- send it to the originating section, not the node.

> we simply

=> we
> so they know they must try another section or try again later.

=> so they can try another section or try again later

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -193,6 +252,28 @@ <h2>Processing a candidate trying to join our section</h2>
   <h2>Resource proof from a destination's point of view</h2>
+  <p>In the previous diagram, we ensured an incoming candidate would only be processed for resource_proof if we are the best fit we know and we are not currently processing another node.<br>
+  This leads us here: to the resource proof.<br>
+  We will first add this node to our peer_list with the sate ResourceProof and send them a RelocateResponse.<br>
```
RelocateResponse => To source section

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -193,6 +252,28 @@ <h2>Processing a candidate trying to join our section</h2>
+  At any time during this process, they may timeout, in which case I will decided to reject them and vote for ParsecPurgeCandidate.<br>
+  This process ends once I reach consensus on either accepting or the candidate (ParsecOnline) or refusing them (ParsecPurgeCandidate).<br>
+  It is possible that both reach consensus, in which case the second one will be discarded in the flow above.
+  </p>
```
It is possible for a node to exit this process having done nothing more than seen ParsecOnline or ParsecPurgeCandidate.

pierrechevalier83 replied:
Well, consensus has to be reached, hasn't it?
Are you trying to say that these have ConsensusMode::Single, so a supermajority is not needed, which is why both may reach consensus?

jeanphilippeD replied:
no, I'm trying to say that consensus can be reached without any input for that node.

jeanphilippeD reviewed (changes requested):
Looks very good.
There were a couple of inaccuracy, so I'll wait to merge it.

pierrechevalier83 commented:
Thanks for the detailed review JP. I've addressed your comments. For a couple of them, I didn't exactly follow your recommendation and explained why in a comment. I'll wait for your confirmation that you're fine with it now :)

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -132,8 +141,59 @@ <h2>Node as a member of a section</h2>
+  </p>
   <h2>Processing a candidate trying to join our section</h2>
+  <p>This flow represents what we do when a candidate contacts us to join our section.<br>
+  The process starts as we receive an ExpectCandidate RPC from this node.<br>
```
> This flow represents what we do when a candidate contacts us to join our section.
> The process starts as we receive an ExpectCandidate RPC from this node.

The flow is from a section contacting us with a candidate. ExpectCandidate is from the section.

jeanphilippeD commented on routing_model/index.html:
```diff
@@ -132,8 +141,59 @@ <h2>Node as a member of a section</h2>
+  </div>
+  <button class="collapsible">ParsecOnline<br>ParsecPurgeCandidate</button>
+  <div class=content>
+    <p>These RPCs come from AcceptAsCandidate.<br>
```
> These RPCs come from AcceptAsCandidate.

These consensused events come from AcceptAsCandidate
